### ISSUE TITLE
mypy duplicates `reveal_type` notes with partial types

### ISSUE THREAD

**Issue by A5rocks** [python/mypy#18511]
**Bug Report**

There's two notes from a single `reveal_type` with a partial type in a loop.

**To Reproduce**

```python
def main() -> None:
    x = None
    while True:
        reveal_type(x)
        x = ""
```

(the outer function is only necessary to make sure a transition to `--local-partial-types` doesn't render this test case useless)

**Expected Behavior**

```
main.py:4: note: Revealed type is "Union[builtins.str, None]"
Success: no issues found in 1 source file
```

**Actual Behavior**

```
main.py:4: note: Revealed type is "None"
main.py:4: note: Revealed type is "Union[builtins.str, None]"
Success: no issues found in 1 source file
```

**Your Environment**

I assume this was caused by https://github.com/python/mypy/pull/18180. Checked on mypy playground.

- Mypy version used: master
- Mypy command-line flags: none
- Mypy configuration options from `mypy.ini` (and other config files): none
- Python version used: 3.12

**Comment by A5rocks** [comment]
I think it's worth it to stop these weird results by preventing early outputs from showing.

That is unless https://github.com/python/mypy/issues/18516 includes redoing partial types.

**Comment by tyralla** [comment]
I guess #18516 would not remove the need to check some loops repeatedly and, hence, would not fix this issue automatically.  Additionally, I suppose, with `--allow-redefinition` disabled, things could generally stay as they are.  But @JukkaL should know this more certainly.
